app/src/main/java/com/tari/android/wallet/infrastructure/backup/WalletBackup.kt

```diff
@@ -0,0 +1,77 @@
+package com.tari.android.wallet.infrastructure.backup
+
+import com.tari.android.wallet.infrastructure.backup.compress.CompressionMethod
+import com.tari.android.wallet.infrastructure.security.encryption.SymmetricEncryptionAlgorithm
+import net.lingala.zip4j.ZipFile
+import net.lingala.zip4j.model.ZipParameters
+import net.lingala.zip4j.model.enums.AesKeyStrength
+import net.lingala.zip4j.model.enums.EncryptionMethod
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileOutputStream
+
+
+interface WalletBackup {
+    fun run(key: CharArray): File
+
+    companion object {
+        fun defaultStrategy(cd: File): WalletBackup = CompressionOnlyBackup.defaultStrategy(cd)
+    }
+}
+
+val BACKUP_FILES = arrayOf("tari_wallet_db.sqlite3")
+
+private class PasswordProtectedZipBackup(private val cd: File) : WalletBackup {
+    override fun run(key: CharArray): File {
+        val zipFile = ZipFile("${cd.absolutePath}/backup_${System.currentTimeMillis()}.zip")
+        val parameters = ZipParameters()
+        if (key.isNotEmpty()) {
+            parameters.isEncryptFiles = true
+            parameters.encryptionMethod = EncryptionMethod.AES
+            parameters.aesKeyStrength = AesKeyStrength.KEY_STRENGTH_256
+            zipFile.setPassword(key)
+        }
+        zipFile.addFiles(BACKUP_FILES.map { File(cd, it) }.toMutableList(), parameters)
+        return zipFile.file
+    }
+}
+
+private class CompositeWalletBackup(
+    private val cd: File,
+    private val encryptionAlgorithm: SymmetricEncryptionAlgorithm,
+    private val compressionMethod: CompressionMethod
+) : WalletBackup {
+
+    override fun run(key: CharArray): File {
+        val backupTime = System.currentTimeMillis()
+        val archive = compressionMethod.compress(
+            "${cd.absolutePath}/temp_$backupTime.zip",
+            BACKUP_FILES.map { File(cd, it) })
+        val encryptedArchiveFile = File(cd, "backup_$backupTime")
+        encryptionAlgorithm.encrypt(
+            key, { FileInputStream(archive) }, { FileOutputStream(encryptedArchiveFile) }
+        )
+        archive.delete()
+        return encryptedArchiveFile
+    }
+
+    companion object {
+        fun defaultStrategy(cd: File) =
+            CompositeWalletBackup(cd, SymmetricEncryptionAlgorithm.aes(), CompressionMethod.zip())
+    }
+
+}
+
+private class CompressionOnlyBackup(
+    private val cd: File,
+    private val compressionMethod: CompressionMethod
+) : WalletBackup {
+    override fun run(key: CharArray): File = compressionMethod.compress(
+        "${cd.absolutePath}/backup_${System.currentTimeMillis()}.zip",
+        BACKUP_FILES.map { File(cd, it) })
+
+    companion object {
+        fun defaultStrategy(cd: File) = CompressionOnlyBackup(cd, CompressionMethod.zip())
```

Comment:
It may make a better API to have these instance in `WalletBackup` companion object.

Reply:
you mean, like having a "compressionOnly(...)" function as a factory method, so that it is possible to instantiate it with `WalletBackup.compressionOnly(args)`?

Reply:
I think so yes, so we get all the options from `WalletBackup` companion. What do you think?

Reply:
I think it will look quite nice